File: Proyecto 2.cpp
//Programa para inventario de productos segun lo que ingrese el usuario
//Realizado por Marlon Armando López Díaz y José Luis Castillo Virula estudiantes de la universidad Mariano Galvez de Guatemala de la faculta de Ingenieria en Sistemas
#include <iostream>
#include <fstream>
#include <string>
#include <limits> 
#include <iomanip>

using namespace std;

const int MAX_PRODUCTOS = 100;

struct Producto {
    string nombre;
    string codigo;
    double precio;
    string proveedor;
    int existencia;
    char estado;
    double descuento;
};

void guardarProductos(Producto productos[], int count) {
    ofstream file("productos.dat", ios::binary);
    file.write(reinterpret_cast<const char*>(productos), count * sizeof(Producto));
    file.close();
}

int cargarProductos(Producto productos[]) {
    ifstream file("productos.dat", ios::binary);
    int count = 0;
    while (file.read(reinterpret_cast<char*>(&productos[count]), sizeof(Producto)) && count < MAX_PRODUCTOS) {
        count++;
    }
    file.close();
    return count;
}

bool contiene(const string& texto, const string& palabra) {
    return texto.find(palabra) != string::npos;
}

bool codigoExistente(Producto productos[], int count, const string& codigo) {
    for (int i = 0; i < count; i++) {
        if (productos[i].codigo == codigo) {
            return true;
        }
    }
    return false;
}

void agregarProducto(Producto productos[], int& count) {
    if (count >= MAX_PRODUCTOS) {
        cout << "No se pueden agregar m?s productos." << endl;
        return;
    }

    Producto nuevo;
    cout << "Ingrese el nombre del producto: ";
    cin.ignore();
    getline (cin, nuevo.nombre);
    cout << "Ingrese el codigo del producto: ";
    cin >> nuevo.codigo;

    if (codigoExistente(productos, count, nuevo.codigo)) {
        cout << "Error: El codigo ya existe." << endl;
        return;
    }

    cout << "Ingrese el precio: ";
    cin >> nuevo.precio;
    cout << "Ingrese el proveedor: ";
    cin.ignore();
    getline (cin, nuevo.proveedor);
    cout << "Ingrese la existencia: ";
    cin >> nuevo.existencia;
    cout << "Ingrese el estado (A/N): ";
    cin >> nuevo.estado;
    cout << "Ingrese el descuento: ";
    cin >> nuevo.descuento;

    productos[count] = nuevo;
    count++;
    guardarProductos(productos, count);
}

void buscarProducto(Producto productos[], int count) {
    string criterio;
    cout << "Ingrese el codigo o nombre del producto a buscar: ";
    cin >> criterio;

    bool encontrado = false;

  
    cout << left << setw(15) << "Nombre"
         << setw(15) << "Codigo"
         << setw(15) << "Precio"
         << setw(15) << "Proveedor"
         << setw(15) << "Existencia"
         << setw(10) << "Estado"
         << setw(10) << "Descuento" << endl;

    for (int i = 0; i < count; i++) {
        if (productos[i].codigo == criterio || contiene(productos[i].nombre, criterio)) {
            cout << left << setw(20) << productos[i].nombre
                 << setw(11) << productos[i].codigo
                 << setw(14)<<productos[i].precio
                 //<< fixed << setprecision(2) << setw(8) << productos[i].precio
                 << setw(15) << productos[i].proveedor
                 << setw(16) << productos[i].existencia
                 << setw(15) << productos[i].estado
                 << setw(15) << productos[i].descuento << endl;
            encontrado = true;
        }
    }

    if (!encontrado) {
        cout << "Producto no encontrado." << endl;
    }
}

void modificarProducto(Producto productos[], int count) {
    string codigo;
    cout << "Ingrese el codigo del producto a modificar: ";
    cin >> codigo;

    for (int i = 0; i < count; i++) {
        if (productos[i].codigo == codigo) {
            cout << "Ingrese el nuevo nombre (actual: " << productos[i].nombre << "): ";
            cin >> productos[i].nombre;
            cin.ignore();
            getline (cin, productos[i].nombre);
            cout << "Ingrese el nuevo precio (actual: " << productos[i].precio << "): ";
            cin >> productos[i].precio;
            cout << "Ingrese el nuevo proveedor (actual: " << productos[i].proveedor << "): ";
            cin >> productos[i].proveedor;
            cin.ignore();
            getline (cin, productos[i].proveedor);
            cout << "Ingrese la nueva existencia (actual: " << productos[i].existencia << "): ";
            cin >> productos[i].existencia;
            cout << "Ingrese el nuevo estado (A/N) (actual: " << productos[i].estado << "): ";
            cin >> productos[i].estado;
            cout << "Ingrese el nuevo descuento (actual: " << productos[i].descuento << "): ";
            cin >> productos[i].descuento;

            guardarProductos(productos, count);
            cout << "Producto modificado." << endl;
            return;
        }
    }
    cout << "Producto no encontrado." << endl;
}

int main() {
    Producto productos[MAX_PRODUCTOS];
    int count = cargarProductos(productos);

    int opcion;
    do {
        cout << "\nMenu:" << endl;
        cout << "1. Agregar producto" << endl;
        cout << "2. Buscar producto" << endl;
        cout << "3. Modificar producto" << endl;
        cout << "4. Salir" << endl;
        cout << "Seleccione una opcion: ";
        cin >> opcion;

        switch (opcion) {
            case 1:
                agregarProducto(productos, count);
                break;
            case 2:
                buscarProducto(productos, count);
                break;
            case 3:
                modificarProducto(productos, count);
                break;
            case 4:
                cout << "Saliendo..." << endl;
                break;
            default:
                cout << "Opcion no valida." << endl;
        }
    } while (opcion != 4);

    return 0;
}

File: proyecto 2.py
#Programa para inventario de productos segun lo que ingrese el usuario.
#Realizado por Marlon Armando Lopez Diaz y Jose Luis Castillo Virula estudiantes de la universidad Mariano Galvez de Guatemala de la faculta de Ingenieria en Sistemas 
import csv
import os

MAX_PRODUCTOS = 100 #Limite de productos que el usuario podra ingresar, sin que se repita el codgio asignado.
# Definicion de la clase de producto
class Producto: #Clase de producto
    def __init__(self, nombre="", codigo="", precio=0.0, proveedor="", existencia=0, estado='A', descuento=0.0):
        #Nombre del producto
        self.nombre = nombre 
        self.codigo = codigo #Codigo unico del producto
        self.precio = precio #Precio del producto
        self.proveedor = proveedor #Proveedor del producto
        self.existencia = existencia #Cantidad disponible
        #Estado del producto (A: Aprobado N: No aprobado)
        self.estado = estado 
        #Descuento aplicable al producto
        self.descuento = descuento 

# Metodo para convertir un objeto Producto a un diccionario
    def to_dict(self):
        return {
            'nombre': self.nombre,
            'codigo': self.codigo,
            'precio': self.precio,
            'proveedor': self.proveedor,
            'existencia': self.existencia,
            'estado': self.estado,
            'descuento': self.descuento,
        }
    # Metodo para crear un objeto Producto a partir de un diccionario
    @classmethod
    def from_dict(cls, data):
        return cls(**data)
# Funcion para guardar productos en un archivo CSV
def guardar_productos_csv(productos):
    with open("productos.csv", mode='w', newline='') as file:
        writer = csv.writer(file)
        # Escribir encabezados
        writer.writerow(['nombre', 'codigo', 'precio', 'proveedor', 'existencia', 'estado', 'descuento'])
        # Escribir los productos
        for producto in productos:
            writer.writerow([producto.nombre, producto.codigo, producto.precio, producto.proveedor, producto.existencia, producto.estado, producto.descuento])

# Funcion para cargar productos desde un archivo CSV
def cargar_productos_csv():
    productos = []
    if os.path.exists("productos.csv"): # Verifica si el archivo existe
        with open("productos.csv", mode='r') as file:
            reader = csv.DictReader(file)
            for row in reader:
                productos.append(Producto(**row))
    return productos
# Funcion para verificar si una cadena contiene otra cadena (criterio de busqueda)
def contiene(texto, palabra):
    return palabra in texto
# Funcion para verificar si el codigo de un producto ya existe
def codigo_existente(productos, codigo):
    return any(producto.codigo == codigo for producto in productos)
# Funcion para agregar un nuevo producto
def agregar_producto(productos):
    if len(productos) >= MAX_PRODUCTOS:
        print("No se pueden agregar más productos.")
        return

    # Solicitar datos del nuevo producto
    nuevo = Producto()
    nuevo.nombre = input("Ingrese el nombre del producto: ")
    nuevo.codigo = input("Ingrese el codigo del producto: ")

   # Verificar si el codigo ya existe
    if codigo_existente(productos, nuevo.codigo):
        print("Error: El codigo ya existe.")
        return
    
    # Ingresar los demas atributos
    nuevo.precio = float(input("Ingrese el precio: "))
    nuevo.proveedor = input("Ingrese el proveedor: ")
    nuevo.existencia = int(input("Ingrese la existencia: "))
    nuevo.estado = input("Ingrese el estado (A/N): ").upper()
    nuevo.descuento = float(input("Ingrese el descuento: "))

    # Agregar el producto a la lista y guardar en CSV
    productos.append(nuevo)
    guardar_productos_csv(productos)

# Funcion para buscar un producto por codigo o nombre
def buscar_producto(productos):
    criterio = input("Ingrese el codigo o nombre del producto a buscar: ")

    encontrado = False
    # Mostrar encabezado de resultados
    print(f"{'Nombre':<25}{'Codigo':<15}{'Precio':<15}{'Proveedor':<15}{'Existencia':<15}{'Estado':<10}{'Descuento':<10}")

    # Buscar el producto
    for producto in productos:
        if producto.codigo == criterio or contiene(producto.nombre, criterio): # Mostrar detalles del producto encontrado
            print(f"{producto.nombre:<25}{producto.codigo:<15}{producto.precio:<15.2f}{producto.proveedor:<15}{producto.existencia:<15}{producto.estado:<10}{producto.descuento:<10.2f}")
            encontrado = True

    if not encontrado:
        print("Producto no encontrado.")

# Funcion para modificar un producto existente
def modificar_producto(productos):
    codigo = input("Ingrese el codigo del producto a modificar: ")

    # Buscar el producto por codigo
    for producto in productos:
        if producto.codigo == codigo:
            # Solicitar nuevos valores para cada atributo
            producto.nombre = input(f"Ingrese el nuevo nombre (actual: {producto.nombre}): ")
            producto.precio = float(input(f"Ingrese el nuevo precio (actual: {producto.precio}): "))
            producto.proveedor = input(f"Ingrese el nuevo proveedor (actual: {producto.proveedor}): ")
            producto.existencia = int(input(f"Ingrese la nueva existencia (actual: {producto.existencia}): "))
            producto.estado = input(f"Ingrese el nuevo estado (A/N) (actual: {producto.estado}): ").upper()
            producto.descuento = float(input(f"Ingrese el nuevo descuento (actual: {producto.descuento}): "))
            # Guardar cambios en el archivo CSV
            guardar_productos_csv(productos)
            print("Producto modificado.")
            return

    print("Producto no encontrado.")

# Funcion principal que controla del programa
def main():
    productos = cargar_productos_csv()

    while True:
        print("\nMenu:")
        print("1. Agregar producto")
        print("2. Buscar producto")
        print("3. Modificar producto")
        print("4. Salir")
        opcion = int(input("Seleccione una opcion: "))

        # Ejecutar la opcion seleccionada
        if opcion == 1:
            agregar_producto(productos)
        elif opcion == 2:
            buscar_producto(productos)
        elif opcion == 3:
            modificar_producto(productos)
        elif opcion == 4:
            print("Saliendo...")
            break
        else:
            print("Opcion no valida.")

# Comprobar si el archivo es ejecutado directamente
if __name__ == "__main__":
    main()
